src/output/plugins/MacOSOutputPlugin.cxx

```diff
@@ -0,0 +1,523 @@
+	buffer_ms = ring_buffer_size / (in_format.mSampleRate * in_format.mBytesPerFrame) * 1000;
+	FormatDebug(macos_output_domain, "Using buffer size of %d ms and %ld bytes", buffer_ms, ring_buffer_size);
+	ring_buffer = new boost::lockfree::spsc_queue<uint8_t>(ring_buffer_size);
+	
+	pcm_export->Open(audio_format.format, audio_format.channels, params);
+	
+	// Register for data request callbacks from the driver and start
+	device.AddIOProc(RenderCallback, this);
+	pause = false;
+}
+
+size_t
+MacOSOutput::Play(const void *chunk, size_t size) {
+	assert(size > 0);
+	
+	if(pause) {
+		pause = false;
+		device.Start();
+	}
+	const auto e = pcm_export->Export({chunk, size});
+	if (e.size == 0)
+		/** The DoP (DSD over PCM) filter converts two frames
+		 *	at a time and ignores the last odd frame; if there
+		 *	was only one frame (e.g. the last frame in the
+		 *	file), the result is empty; to avoid an endless
+		 *	loop, bail out here, and pretend the one frame has
+		 *	been played
+		 */
+		return size;
+
+	size_t bytes_written = ring_buffer->push((const uint8_t *)e.data, e.size);
+	return pcm_export->CalcSourceSize(bytes_written);
+}
+
+std::chrono::steady_clock::duration
+MacOSOutput::Delay() const noexcept {
+	if(pause)
+		return std::chrono::seconds(1);
+	// Wait for half the buffer size in case the buffer is full
+	return ring_buffer->write_available() ? std::chrono::steady_clock::duration::zero() : std::chrono::milliseconds(buffer_ms / 2);
+}
+	
+bool
+MacOSOutput::Pause() {
+	if(!pause) {
+		pause = true;
+		device.Stop();
+	}
+	return true;
+}
+	
+int
+macos_output_get_volume(MacOSOutput &output) {
+	return output.GetVolume();
+}
+
+void
+macos_output_set_volume(MacOSOutput &output, unsigned new_volume) {
+	return output.SetVolume(new_volume);
+}
+	
+OSStatus
+MacOSOutput::RenderCallback(gcc_unused AudioObjectID inDevice, gcc_unused const AudioTimeStamp* inNow, gcc_unused const AudioBufferList* inInputData, gcc_unused const AudioTimeStamp* inInputTime, AudioBufferList* outOutputData, gcc_unused const AudioTimeStamp* inOutputTime, void* inClientData) {
+	MacOSOutput *oo = (MacOSOutput*)inClientData;
+	UInt32 requested = outOutputData->mBuffers[oo->device.GetStreamIdx()].mDataByteSize;
+	/** Frames are the same for both MPD input format and CoreAudio
+	 *	output format as the sample rates are matching.
+	 */
+	UInt32 frames = requested / oo->out_format.mBytesPerFrame;
+	// Number of bytes to pop from ring_buffer (MPD format framesize times number of frames)
+	UInt32 in_bytes = std::min(frames * oo->in_format.mBytesPerFrame, (UInt32) oo->ring_buffer->read_available());
+	UInt32 available_frames = in_bytes / oo->in_format.mBytesPerFrame;
+	
+	if(available_frames < frames)
+		FormatDebug(macos_output_domain, "Frames available (%d) less than requested (%d) by device.", available_frames, frames);
+	
+	// Usage of converter
```

Comment:
I don't like the idea of using a converter. If you let AudioUnit to handle the stream, it will automatically convert the format for you. The conversion is transparent --- if your device's output format is identical to the file, it will just copy the bytes. 

I don't know if it's a better approach here to do the conversion by yourself and manipulate various objects (such as virtual format). The original code can do pretty much everything you are trying to achieve here  --- #271  could be easily solved by apply a patch that's similar to the reversion of https://github.com/cmus/cmus/pull/456